File: CPP/1-Creationals/Prototype/Prototype.cpp
#include <iostream>
#include <vector>

// Prototype base class
class Filter
{
public:
    virtual ~Filter() {}
    virtual Filter* clone() = 0;
    virtual void applyFilter() = 0;
};

// Factory method used to allocate the prototypes
class Factory
{
public:
    static Filter *makeFilter( int filter_index );

private:
    // This factory has 3 types of prototype
    static Filter *s_prototypes[3];
};

class GaussianBlur : public Filter
{
public:
    /*virtual*/ Filter *clone()
    {
        return new GaussianBlur();
    }
    /*virtual*/ void applyFilter()
    {
        std::cout << "GaussianBlur: applying filter...\n";
    }
};

class Pixelate : public Filter
{
public:
    /*virtual*/ Filter *clone()
    {
        return new Pixelate();
    }
    /*virtual*/ void applyFilter()
    {
        std::cout << "Pixelate: applying filter...\n";
    }
};

class Threshold : public Filter
{
public:
    /*virtual*/ Filter *clone()
    {
        return new Threshold();
    }
    /*virtual*/ void applyFilter()
    {
        std::cout << "Threshold: applying filter...\n";
    }
};

// Initialize the prototypes' list of Filters
Filter *Factory::s_prototypes[] = {
    new GaussianBlur(), new Pixelate(), new Threshold()
};

// Implementation of the Factory method used to create cloned prototypes objects
Filter *Factory::makeFilter( int filter_index )
{
    return s_prototypes[filter_index]->clone();
}

// Use the prototype pattern
int main()
{
    std::vector<Filter *> filters;
    int filter_index;

    // Create some filters
    // 0 => GaussianBlur, 1 => Pixelate, 2 => Threshold
    filters.push_back( Factory::makeFilter( 0 ) );
    filters.push_back( Factory::makeFilter( 2 ) );
    filters.push_back( Factory::makeFilter( 1 ) );

    // Use all the created filters
    for (int i=0; i < filters.size(); ++i)
    {
        filters[i]->applyFilter();
    }
    // free the allocated memory
    for (int i=0; i < filters.size(); ++i)
    {
        delete filters[i];
    }

    return 0;
}

//-----------------------------------------------------------
/*
Output:

GaussianBlur: applying filter...
Threshold: applying filter...
Pixelate: applying filter...

*/

File: Python/1-Creationals/Prototype/Prototype.py
#!/usr/bin/env python3

from abc import ABC, abstractmethod


class Filter(ABC):
    """
    Base class for the prototype pattern
    """
    def __init__(self):
        pass

    @abstractmethod
    def clone(self):
        """
        Method used to generate the cloned instances
        :return:
        """
        pass

    @abstractmethod
    def apply_filter(self):
        """
        Method used to demonstrate the execution of different filters
        :return: None
        """
        pass


class GaussianBlur(Filter):
    def clone(self):
        return GaussianBlur()

    def apply_filter(self):
        print("GaussianBlur: applying filter...")


class Pixelate(Filter):
    def clone(self):
        return Pixelate()

    def apply_filter(self):
        print("Pixelate: applying filter...")


class Threshold(Filter):
    def clone(self):
        return Threshold()

    def apply_filter(self):
        print("Threshold: applying filter...")


class Factory(ABC):
    """
    Factory used to create objects from the corresponding prototype
    """
    prototypes = [GaussianBlur(), Pixelate(), Threshold()]

    @staticmethod
    def make_filter(index):
        return Factory.prototypes[index].clone()


def main():
    # Create some filters
    # 0 => GaussianBlur, 1 => Pixelate, 2 => Threshold
    filters = [Factory.make_filter(0), Factory.make_filter(2), Factory.make_filter(1)]

    for f in filters:
        f.apply_filter()


if __name__ == "__main__":
    main()


"""
Output:

GaussianBlur: applying filter...
Threshold: applying filter...
Pixelate: applying filter...

"""
